File: nodejs-typescript/src/plan.ts
import { createHash } from "node:crypto";
import { readFile } from "node:fs/promises";
import { fileURLToPath } from "node:url";

import type { PlanCase, PlanSnapshot, RequestPlan, Scenario } from "./types.js";

const PLAN_PATH = fileURLToPath(
  new URL("../../fixtures/request-plan.json", import.meta.url),
);

const EXPECTED_CASES: ReadonlyArray<
  Readonly<Pick<PlanCase, "id" | "scenario" | "model" | "thinking" | "max_tokens">>
> = [
  {
    id: "node-ordinary-chat-v4-flash",
    scenario: "ordinary_chat",
    model: "deepseek-v4-flash",
    thinking: "disabled",
    max_tokens: 32,
  },
  {
    id: "node-stream-v4-flash",
    scenario: "streaming",
    model: "deepseek-v4-flash",
    thinking: "disabled",
    max_tokens: 32,
  },
  {
    id: "node-json-mode-v4-flash",
    scenario: "json_output",
    model: "deepseek-v4-flash",
    thinking: "disabled",
    max_tokens: 64,
  },
  {
    id: "node-tool-initial-v4-flash",
    scenario: "tool_initial",
    model: "deepseek-v4-flash",
    thinking: "disabled",
    max_tokens: 64,
  },
  {
    id: "node-tool-continuation-v4-flash",
    scenario: "tool_continuation",
    model: "deepseek-v4-flash",
    thinking: "disabled",
    max_tokens: 48,
  },
  {
    id: "node-thinking-v4-pro",
    scenario: "thinking",
    model: "deepseek-v4-pro",
    thinking: "enabled",
    max_tokens: 96,
  },
  {
    id: "node-alias-deepseek-chat",
    scenario: "alias_probe",
    model: "deepseek-chat",
    thinking: "disabled",
    max_tokens: 16,
  },
  {
    id: "node-alias-deepseek-reasoner",
    scenario: "alias_probe",
    model: "deepseek-reasoner",
    thinking: "enabled",
    max_tokens: 32,
  },
  {
    id: "node-invalid-model-error",
    scenario: "invalid_model",
    model: "deepseek-does-not-exist",
    thinking: "disabled",
    max_tokens: 16,
  },
];

const SCENARIOS = new Set<Scenario>([
  "ordinary_chat",
  "streaming",
  "json_output",
  "tool_initial",
  "tool_continuation",
  "thinking",
  "alias_probe",
  "invalid_model",
]);

function sameExpectedCase(actual: PlanCase, expected: (typeof EXPECTED_CASES)[number]): boolean {
  return (
    actual.id === expected.id &&
    actual.scenario === expected.scenario &&
    actual.model === expected.model &&
    actual.thinking === expected.thinking &&
    actual.max_tokens === expected.max_tokens
  );
}

export function validatePlan(value: unknown): asserts value is RequestPlan {
  if (!value || typeof value !== "object") {
    throw new Error("The request plan must be an object.");
  }
  const plan = value as Partial<RequestPlan>;
  if (!Array.isArray(plan.cases)) {
    throw new Error("The request plan must contain cases.");
  }
  if (
    plan.schema_version !== 1 ||
    plan.status !== "not_run" ||
    plan.provider_origin !== "https://api.deepseek.com"
  ) {
    throw new Error("The frozen request plan header changed.");
  }
  if (
    plan.provider_request_cap !== 9 ||
    plan.planned_provider_requests !== 9 ||
    plan.cases.length !== 9
  ) {
    throw new Error("The provider request cap must remain exactly nine.");
  }
  if (
    plan.concurrency !== 1 ||
    plan.automatic_retries !== 0 ||
    plan.default_timeout_ms !== 30000
  ) {
    throw new Error("Concurrency, retry, or timeout controls changed.");
  }
  if (
    plan.versions?.openai !== "6.49.0" ||
    plan.versions.typescript !== "7.0.2" ||
    plan.versions["@types/node"] !== "24.13.3" ||
    plan.versions.node_minimum !== "20"
  ) {
    throw new Error("A frozen dependency pin changed.");
  }
  if (
    !plan.evidence_policy ||
    Object.values(plan.evidence_policy).some((allowed) => allowed !== false)
  ) {
    throw new Error("Every persistence policy must remain false.");
  }

  const ids = new Set<string>();
  for (let index = 0; index < plan.cases.length; index += 1) {
    const actual = plan.cases[index];
    const expected = EXPECTED_CASES[index];
    if (!actual || !expected || !sameExpectedCase(actual, expected)) {
      throw new Error(`Frozen case ${index + 1} changed.`);
    }
    if (
      actual.sequence !== index + 1 ||
      !SCENARIOS.has(actual.scenario) ||
      ids.has(actual.id) ||
      typeof actual.acceptance !== "string" ||
      actual.acceptance.length < 20
    ) {
      throw new Error(`Frozen case ${index + 1} is invalid.`);
    }
    if (
      actual.reasoning_effort !== (actual.thinking === "enabled" ? "high" : null)
    ) {
      throw new Error(`Frozen case ${index + 1} has invalid reasoning controls.`);
    }
    ids.add(actual.id);
  }
}

export async function loadPlan(path = PLAN_PATH): Promise<PlanSnapshot> {
  const encoded = await readFile(path, "utf8");
  const parsed: unknown = JSON.parse(encoded);
  validatePlan(parsed);
  return {
    plan: parsed,
    sha256: createHash("sha256").update(encoded, "utf8").digest("hex"),
  };
}

export const EXPECTED_PROVIDER_REQUEST_CAP = 9;

File: langchain-integration/src/plan.py
"""Frozen cross-runtime provider request-plan loading and validation."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


ROOT = Path(__file__).resolve().parents[1]
PLAN_PATH = ROOT / "fixtures" / "request-plan.json"
EXPECTED_PYTHON_VERSIONS = {
    "langchain": "1.3.14",
    "langchain-deepseek": "1.1.0",
    "langchain-openai": "1.4.1",
    "langchain-core": "1.5.1",
    "openai": "2.48.0",
    "pydantic": "2.13.4",
}
EXPECTED_JAVASCRIPT_VERSIONS = {
    "@langchain/deepseek": "1.1.5",
    "@langchain/core": "1.2.3",
    "zod": "4.4.3",
    "node_minimum": "20",
}


def load_plan() -> dict[str, Any]:
    plan = json.loads(PLAN_PATH.read_text(encoding="utf-8"))
    validate_plan(plan)
    return plan


def validate_plan(plan: dict[str, Any]) -> None:
    cases = plan.get("cases")
    if not isinstance(cases, list):
        raise ValueError("Plan must contain a cases array.")
    if plan.get("status") != "not_run":
        raise ValueError("The frozen plan must remain marked not_run.")
    if plan.get("provider_origin") != "https://api.deepseek.com":
        raise ValueError("The provider origin must remain exact.")
    if plan.get("provider_request_cap") != 16:
        raise ValueError("The provider request cap must remain exactly 16.")
    if plan.get("planned_provider_requests") != 16 or len(cases) != 16:
        raise ValueError("The plan must contain exactly 16 provider cases.")
    if plan.get("concurrency") != 1 or plan.get("automatic_retries") != 0:
        raise ValueError("The live plan must use concurrency 1 and zero retries.")

    versions = plan.get("versions", {})
    if versions.get("python") != EXPECTED_PYTHON_VERSIONS:
        raise ValueError("Unexpected Python dependency pins.")
    if versions.get("javascript") != EXPECTED_JAVASCRIPT_VERSIONS:
        raise ValueError("Unexpected JavaScript dependency pins.")

    sequences = [case.get("sequence") for case in cases]
    if sequences != list(range(1, 17)):
        raise ValueError("Case sequences must be contiguous and ordered.")
    ids = [case.get("id") for case in cases]
    if len(ids) != len(set(ids)):
        raise ValueError("Case IDs must be unique.")

    python_cases = [case for case in cases if case.get("runtime") == "python"]
    javascript_cases = [case for case in cases if case.get("runtime") == "javascript"]
    if len(python_cases) != 11 or len(javascript_cases) != 5:
        raise ValueError("The plan must contain 11 Python and five JavaScript cases.")

    scenario_counts = {
        scenario: sum(case.get("scenario") == scenario for case in cases)
        for scenario in {
            "invoke",
            "stream",
            "structured_output",
            "tool_initial",
            "tool_continuation",
            "local_context_rag",
            "alias_probe",
            "invalid_model",
        }
    }
    expected = {
        "invoke": 4,
        "stream": 3,
        "structured_output": 2,
        "tool_initial": 1,
        "tool_continuation": 1,
        "local_context_rag": 1,
        "alias_probe": 2,
        "invalid_model": 2,
    }
    if scenario_counts != expected:
        raise ValueError(f"Unexpected scenario matrix: {scenario_counts}")

    aliases = {
        case["model"] for case in cases if case.get("scenario") == "alias_probe"
    }
    if aliases != {"deepseek-chat", "deepseek-reasoner"}:
        raise ValueError("The plan must contain the two frozen alias probes.")

    for case in cases:
        if case.get("thinking") not in {"enabled", "disabled"}:
            raise ValueError(f"Thinking must be explicit for {case.get('id')}.")
        max_tokens = case.get("max_tokens")
        if not isinstance(max_tokens, int) or not 16 <= max_tokens <= 96:
            raise ValueError(f"Invalid token cap for {case.get('id')}.")


def cases_for_runtime(plan: dict[str, Any], runtime: str) -> list[dict[str, Any]]:
    return [case for case in plan["cases"] if case["runtime"] == runtime]
